File: nodejs-backend/src/services/scrim.service.js
/**
 * Scrim Service
 * Handles scrim-related business logic
 */

const { Scrim, Team } = require('../models');

class ScrimService {
  /**
   * Get all scrims
   */
  async getAllScrims() {
    return await Scrim.findAll({
      include: [{
        model: Team,
        as: 'team',
        attributes: ['id', 'name', 'game']
      }],
      order: [['date', 'DESC']]
    });
  }

  /**
   * Get scrim by ID
   */
  async getScrimById(id) {
    return await Scrim.findByPk(id, {
      include: [{
        model: Team,
        as: 'team',
        attributes: ['id', 'name', 'game']
      }]
    });
  }

  /**
   * Get scrims by status
   */
  async getScrimsByStatus(status) {
    return await Scrim.findAll({
      where: { status },
      include: [{
        model: Team,
        as: 'team',
        attributes: ['id', 'name', 'game']
      }],
      order: [['date', 'DESC']]
    });
  }

  /**
   * Get scrims by team ID
   */
  async getScrimsByTeamId(teamId) {
    return await Scrim.findAll({
      where: { team_id: teamId },
      include: [{
        model: Team,
        as: 'team',
        attributes: ['id', 'name', 'game']
      }],
      order: [['date', 'DESC']]
    });
  }

  /**
   * Create new scrim
   */
  async createScrim(scrimData) {
    const scrim = await Scrim.create({
      team_id: scrimData.teamId,
      opponent: scrimData.opponent,
      description: scrimData.description,
      date: scrimData.date,
      status: scrimData.status || 'PENDING',
      game: scrimData.game,
      notes: scrimData.notes
    });

    return await this.getScrimById(scrim.id);
  }

  /**
   * Update scrim
   */
  async updateScrim(id, scrimData) {
    const scrim = await Scrim.findByPk(id);
    if (!scrim) {
      throw new Error('Scrim not found');
    }

    if (scrimData.opponent !== undefined) scrim.opponent = scrimData.opponent;
    if (scrimData.description !== undefined) scrim.description = scrimData.description;
    if (scrimData.date !== undefined) scrim.date = scrimData.date;
    if (scrimData.status !== undefined) scrim.status = scrimData.status;
    if (scrimData.game !== undefined) scrim.game = scrimData.game;
    if (scrimData.notes !== undefined) scrim.notes = scrimData.notes;

    await scrim.save();
    return await this.getScrimById(scrim.id);
  }

  /**
   * Delete scrim
   */
  async deleteScrim(id) {
    const scrim = await Scrim.findByPk(id);
    if (!scrim) {
      throw new Error('Scrim not found');
    }

    await scrim.destroy();
    return true;
  }
}

module.exports = new ScrimService();

File: flutter_app/lib/services/scrim_service.dart
import '../models/scrim.dart';
import 'api_service.dart';

/// Scrim service for managing scrims
class ScrimService {
  final ApiService _api = ApiService();

  /// Get all scrims
  Future<List<Scrim>> getAllScrims() async {
    final response = await _api.get('/api/scrims');
    return (response as List).map((json) => Scrim.fromJson(json)).toList();
  }

  /// Get scrim by ID
  Future<Scrim> getScrimById(int id) async {
    final response = await _api.get('/api/scrims/$id');
    return Scrim.fromJson(response);
  }

  /// Get scrims by status
  Future<List<Scrim>> getScrimsByStatus(String status) async {
    final response = await _api.get('/api/scrims/status/$status');
    return (response as List).map((json) => Scrim.fromJson(json)).toList();
  }

  /// Get scrims by team ID
  Future<List<Scrim>> getScrimsByTeamId(int teamId) async {
    final response = await _api.get('/api/scrims/team/$teamId');
    return (response as List).map((json) => Scrim.fromJson(json)).toList();
  }

  /// Create scrim (admin only)
  Future<Scrim> createScrim(Scrim scrim) async {
    final response = await _api.post('/api/scrims', body: scrim.toJson());
    return Scrim.fromJson(response);
  }

  /// Update scrim (admin only)
  Future<Scrim> updateScrim(int id, Scrim scrim) async {
    final response = await _api.put('/api/scrims/$id', body: scrim.toJson());
    return Scrim.fromJson(response);
  }

  /// Delete scrim (admin only)
  Future<void> deleteScrim(int id) async {
    await _api.delete('/api/scrims/$id');
  }
}
